infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogCompaction.java

```diff
@@ -0,0 +1,63 @@
+package org.corfudb.infrastructure.log;
+
+import com.codahale.metrics.Counter;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.util.MetricsUtils;
+
+import java.util.concurrent.*;
+
+/**
+ * Scheduled log compaction manager.
+ * Execute log compaction safely (prevent scheduled executor corruption if the task throws an exception)
+ * https://stackoverflow.com/questions/6894595/scheduledexecutorservice-exception-handling
+ */
+@Slf4j
+public class StreamLogCompaction {
+    private static final String METRIC_NAME = StreamLogCompaction.class.getName();
+
+    private final ThreadFactory threadFactory = new ThreadFactoryBuilder()
+            .setDaemon(true)
+            .setNameFormat("LogUnit-Maintenance-%d")
+            .build();
+
+    /**
+     * How many times log compaction executed
+     */
+    private final Counter gc;
+
+    /**
+     * A scheduler, which is used to schedule periodic garbage collection.
+     */
+    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor(threadFactory);
+    private final ScheduledFuture<?> compactor;
+
+    public StreamLogCompaction(StreamLog streamLog, long initialDelay, long period, TimeUnit timeUnit) {
+        gc = MetricsUtils.metrics.counter(String.format("%s-stream[%d]", METRIC_NAME, streamLog.hashCode()));
+
+        Runnable task = () -> {
+            gc.inc();
+            log.debug("Start log compaction. Counter: {}", gc.getCount());
+            try {
+                streamLog.compact();
+            } catch (Throwable ex) {
+                log.error("Can't compact stream log. Counter: {}", gc.getCount());
```

Comment:
Why happens if the thread is interrupted?
Also, if an exception happens here, then we might need to bail.

Reply:
Any exception just breaks the task and scheduledPool can be stopped if the shutdown method is called.

I believe we can't do anything in the task if an exception happens, we just write that fact in a log and try to compact the log after a timeout. If you have any ideas what we can do in this case, please let me know.